beacon-chain/slasher/chunks.go

```diff
@@ -375,10 +389,89 @@ func (m *MaxSpanChunksSlice) Update(
 	// continue to the next chunk to update it.
-	keepGoing = epochInChunk <= currentEpoch
+	keepGoing = epochInChunk <= opts.currentEpoch
+	return
+}
+
+// StartEpoch given a source epoch and current epoch, determines the start epoch of
+// a min span chunk for use in chunk updates. To compute this value, we look at the difference between
+// H = HistoryLength and the current epoch. Then, we check if the source epoch > difference. If so,
+// then the start epoch is source epoch - 1. Otherwise, we return the caller a boolean signifying
+// the input argumets are invalid for the chunk and the start epoch does not exist.
+func (m *MinSpanChunksSlice) StartEpoch(
+	sourceEpoch, currentEpoch types.Epoch,
+) (epoch types.Epoch, exists bool) {
+	// Given min span chunks are used for detecting surrounding votes, we have no need
+	// for a start epoch of the chunk if the source epoch is 0 in the input arguments.
+	// To further clarify, min span chunks are updated in reverse order [a, b, c, d] where
+	// if the start epoch is d, then we down the chunk updating everything from d, c, b, to
+	// a. If the source epoch is 0, this would correspond to a, which means there is nothing
+	// more to update.
+	if sourceEpoch == 0 {
+		return
+	}
+	var difference uint64
+	if uint64(currentEpoch) > m.params.historyLength {
+		difference = uint64(currentEpoch) - m.params.historyLength
+	}
+	if uint64(sourceEpoch) <= difference {
+		return
+	}
+	epoch = sourceEpoch.Sub(1)
+	exists = true
+	return
+}
+
+// StartEpoch given a source epoch and current epoch, determines the start epoch of
+// a max span chunk for use in chunk updates. The source epoch cannot be >= the current epoch.
+func (m *MaxSpanChunksSlice) StartEpoch(
+	sourceEpoch, currentEpoch types.Epoch,
+) (epoch types.Epoch, exists bool) {
+	if sourceEpoch >= currentEpoch {
+		return
+	}
+	// Given max spans is a list of max targets for source epochs, the precondition is that
+	// every attestation's source epoch must be < than its target epoch. So the start epoch
+	// for updates is given as source epoch + 1.
+	epoch = sourceEpoch.Add(1)
+	exists = true
 	return
 }
 
+// NextChunkStartEpoch given an epoch, determines the start epoch of the next chunk. For min
+// span chunks, this will be the last epoch of chunk index = (current chunk - 1). For example:
+//
+//                       chunk0     chunk1     chunk2
+//                         |          |          |
+//  max_spans_val_i = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]
+//
+// If C = ChunkSize is 3 epochs per chunk, and we input start epoch of chunk 1 which is 3. The next start
+// epoch is the last epoch of chunk 0, which is epoch 2. This is computed as:
+//
+//  (start_epoch / C) * (C - 1)
+//  (3 / 3) * (3 - 1) = 1 * 2 = 2
+//
+func (m *MinSpanChunksSlice) NextChunkStartEpoch(startEpoch types.Epoch) types.Epoch {
+	return types.Epoch(uint64(startEpoch)/m.params.chunkSize*m.params.chunkSize - 1)
```

Comment:
```suggestion
	return types.Epoch((uint64(startEpoch)/m.params.chunkSize)*(m.params.chunkSize - 1))
```
Add parenthesis to match godoc description.

This actually changes the logic. Please double check that the godoc and implementation match and are correct.